starknet_py/utils/typed_data.py

```diff
@@ -360,6 +398,54 @@ def encode_bool(value: Union[bool, str, int]) -> int:
     raise ValueError(f"Expected boolean value, got [{value}].")
 
 
+def is_digit_string(s: str, signed=False) -> bool:
+    if signed:
+        return bool(re.fullmatch(r"-?\d+", s))
+    return bool(re.fullmatch(r"\d+", s))
+
+
+def encode_u128(value: Union[str, int]) -> int:
+    def is_in_range(n: int):
+        return 0 <= n < 2**128
+
+    if isinstance(value, str) and value.startswith("0x"):
+        int_value = int(value, 16)
+    elif isinstance(value, str) and is_digit_string(value):
+        int_value = int(value)
+    elif isinstance(value, int):
+        int_value = value
+    else:
+        raise ValueError(f"Value [{value}] is not a valid number.")
+
+    if is_in_range(int_value):
+        return int_value
+    raise ValueError(f"Value [{value}] is out of range for '{BasicType.U128}'.")
+
+
+def encode_i128(value: Union[str, int]) -> int:
+    def is_in_range(n: int):
+        return (n < 2**127) or (n >= (FIELD_PRIME - (2**127)))
+
+    if isinstance(value, str) and value.startswith("0x"):
+        int_value = int(value, 16)
+    elif isinstance(value, str) and is_digit_string(value, True):
+        int_value = int(value)
+    elif isinstance(value, int):
+        int_value = value
+    else:
+        raise ValueError(f"Value [{value}] is not a valid number.")
+
+    if abs(int_value) >= FIELD_PRIME:
+        raise ValueError(
+            f"Values outside the range (-FIELD_PRIME, FIELD_PRIME) are not allowed, [{value}] given."
+        )
+    int_value %= FIELD_PRIME
```

Comment:
In case of **signed** number, integer value need to be in range `(-P, P)`, because of they way how negative numbers are calculated. Here's an official [docs reference](https://docs.starknet.io/architecture-and-concepts/smart-contracts/serialization-of-cairo-types/#data_types_greater_than_252_bits).

Also, as other reference, here's a [similar implementation](https://github.com/software-mansion/starknet-jvm/blob/main/lib/src/main/kotlin/com/swmansion/starknet/data/types/Felt.kt#L117-L122) in JVM SDK.